### PR TITLE
[ADAM-536] Add support for scala 2.11

### LINKED ISSUE #536: scala 2.11
I see that Spark 1.20 supports scala 2.11, it would be great if adam would also move to it.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #536. Doesn't _quite_ work yet; still working to resolve this issue with the scala 2.11 build:

```
[ERROR] /Users/fnothaft/IdeaProjects/adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala:94: diverging implicit expansion for type ((org.bdgenomics.adam.models.ReferenceRegion, Int)) => Comparable[(org.bdgenomics.adam.models.ReferenceRegion, Int)]
starting with method orderingToOrdered in object Ordered
[ERROR]       keyedLeft.repartitionAndSortWithinPartitions(ManualRegionPartitioner(bins.value.numBins))
[ERROR]       ^
[ERROR] /Users/fnothaft/IdeaProjects/adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala:94: value repartitionAndSortWithinPartitions is not a member of org.apache.spark.rdd.RDD[((org.bdgenomics.adam.models.ReferenceRegion, Int), T)]
[ERROR]       keyedLeft.repartitionAndSortWithinPartitions(ManualRegionPartitioner(bins.value.numBins))
[ERROR]                 ^
[ERROR] /Users/fnothaft/IdeaProjects/adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala:96: diverging implicit expansion for type ((org.bdgenomics.adam.models.ReferenceRegion, Int)) => Comparable[(org.bdgenomics.adam.models.ReferenceRegion, Int)]
starting with method orderingToOrdered in object Ordered
[ERROR]       keyedRight.repartitionAndSortWithinPartitions(ManualRegionPartitioner(bins.value.numBins))
[ERROR]       ^
[ERROR] /Users/fnothaft/IdeaProjects/adam/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala:96: value repartitionAndSortWithinPartitions is not a member of org.apache.spark.rdd.RDD[((org.bdgenomics.adam.models.ReferenceRegion, Int), U)]
[ERROR]       keyedRight.repartitionAndSortWithinPartitions(ManualRegionPartitioner(bins.value.numBins))
```

@laserson I was trying to figure out where the ordering for `(ReferenceRegion, Int)` was defined/is coming in from, but I wasn't able to grok that from the `ShuffleRegionJoin` code. Do you know where that is coming from/would you be able to take a look?

### CODE REVIEW

**Comment by fnothaft** [comment]
Also, it looks like this problem is related to a [regression](https://issues.scala-lang.org/browse/SI-8541) in Scala 2.11, but I believe we should be able to WAR the regression.

**Comment by laserson** [comment]
Just submitted a PR against your branch that fixes the Ordering problem.  Mo' problems to solve, though...

**Comment by laserson** [comment]
(Also, I'm not sure where the Ordering was coming from to begin with, at it appears Tuples don't have default Orderings.)

**Comment by fnothaft** [comment]
Thanks @laserson! I was confused as well...

**Comment by fnothaft** [comment]
Jenkins, retest this please.

**Comment by massie** [comment]
Jenkins, test this please.

**Comment by massie** [comment]
This approach looks good to me. I'd like to hear from @heuermh too since he's strong with the maven fu.

File: pom.xml [hunk 1]
```diff
@@ -483,6 +487,13 @@
 
   <profiles>
     <profile>
+      <id>scala-2.11</id>
+      <properties>
+        <scala.version>2.11.2</scala.version>
```
Comment (massie) [inline pom.xml]: We are up to 2.11.5 now and `brew` has 2.11.4. We might want to bump this up before we commit it?

**Comment by massie** [comment]
It looks like there are a few places where 2.10 is still hard-coded, e.g.

```
$ find . -name "pom.xml" | xargs grep "2.10"
./adam-cli/pom.xml:      <artifactId>scalate-core_2.10</artifactId>
./adam-cli/pom.xml:      <artifactId>scalatra-json_2.10</artifactId>
./adam-cli/pom.xml:      <artifactId>scalatra_2.10</artifactId>
./pom.xml:    <scala.version>2.10.4</scala.version>
./pom.xml:    <scala.artifact.suffix>2.10</scala.artifact.suffix>
./pom.xml:        <artifactId>scalate-core_2.10</artifactId>
./pom.xml:        <artifactId>scalatra-json_2.10</artifactId>
./pom.xml:        <artifactId>scalatra_2.10</artifactId>
```
we should probably update this to use `$scala.artifact.suffix`

**Comment by massie** [comment]
This doesn't build on my Mac, e.g.

```
$ mvn -P scala-2.11 clean package
```
Throws the following error:
```
[ERROR] /Users/matt/workspace/berkeley/adam/adam-core/src/main/scala/org/bdgenomics/adam/models/VariantContext.scala:84: in class VariantContext, multiple overloaded alternatives of constructor VariantContext define default arguments.
[ERROR] class VariantContext(
[ERROR]       ^
[WARNING] 5 warnings found
[ERROR] one error found
```
Things build fine when I remove the `scala-2.11` profile.

Are others seeing the same issue?

**Comment by fnothaft** [comment]
Let me take a look; this built on my end when I opened the PR, not sure if there's been a change since then that'd break it...

**Comment by fnothaft** [comment]
Er, duh! The 2.10 would build; 2.11 wouldn't build because the error in the original PR post wasn't fixed yet. I'll take a look at this locally now.

**Comment by fnothaft** [comment]
This should work for Scala 2.11 now. Can someone test on their own machine?

**Comment by heuermh** [comment]
Also fails for me, see https://gist.github.com/heuermh/14f7b33def990f4eb7a2

**Comment by fnothaft** [comment]
Ah! Looking at this again...

**Comment by fnothaft** [comment]
Closing this. Will open a new PR tomorrow that builds on https://github.com/bigdatagenomics/utils/pull/36.

**Pull request closed without merging**